Improve `NumericInput` behavior for valid numbers and key-press

AnuragVasanwala opened the pull request:
## Context

The goal of the PR is to fix `NumberInput` bugs:
- [x] Allow numeric characters of fixed length (not more than character length of `max`). (https://github.com/GoogleForCreators/web-stories-wp/issues/13238)
- [x] Remove unnecessary padding Zero (https://github.com/GoogleForCreators/web-stories-wp/issues/13239)
- [x] Update numeric value when inputted via numeric keys. (https://github.com/GoogleForCreators/web-stories-wp/issues/13240)
- [x] Support for full key up-down values between `min` and `max`.

## Summary

Fixes bug with `NumericInput` component. Detailed changes are written in the "User-facing changes" section.

## Relevant Technical Choices

Not applicable.

## To-do

Not applicable.

## User-facing changes

| Scenario  | Image / Animated GIF
| ------------- | ------------- |
| Allow fixed length of numeric characters (https://github.com/GoogleForCreators/web-stories-wp/issues/13238) | ![Fixed-Length-of-Numbers](https://github.com/GoogleForCreators/web-stories-wp/assets/75766877/053bb8d0-1138-4fd4-981c-028e5313667a) |
| Remove unnecessary padding (previously padding was not removed even if length of characters exceeds maximum character length of `max` value) (https://github.com/GoogleForCreators/web-stories-wp/issues/13239) | <img width="362" alt="image" src="https://github.com/GoogleForCreators/web-stories-wp/assets/75766877/45596cf6-9494-4883-99fc-051b74513af3"> |
| Update time when `NumericInput` is updated via keys using number keys (https://github.com/GoogleForCreators/web-stories-wp/issues/13240) | ![Update-time-with-Number-keys](https://github.com/GoogleForCreators/web-stories-wp/assets/75766877/617544c0-e1f6-47f6-ad45-0a8b1782483b) |
| Support for full key up-down values between `min` and `max` (previously zero was not allowed in `minutes`) | ![Full-key-up-down-support](https://github.com/GoogleForCreators/web-stories-wp/assets/75766877/d2fc4db5-3471-4499-b5ee-960c4fc7fdf3) |

## Testing Instructions

Refer related issues this PR fixes for testing instruction.

## Reviews

### Does this PR have a security-related impact?

No.

### Does this PR change what data or activity we track or use?

No.

### Does this PR have a legal-related impact?

No.

## Checklist

<!-- Check these after PR creation -->

- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/googleforcreators/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [ ] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/googleforcreators/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---
Fixes #13238, Fixes #13239 and Fixes #13240

AnuragVasanwala committed:
🐛 Fix: `NumericInput` Component for valid numbers and key-press

AnuragVasanwala committed:
Merge branch 'main' into 13238-13239-13240-numeric-field-bug

AnuragVasanwala committed:
Merge branch 'main' into 13238-13239-13240-numeric-field-bug

AnuragVasanwala committed:
Merge branch 'main' into 13238-13239-13240-numeric-field-bug

AnuragVasanwala committed:
Merge branch 'main' into 13238-13239-13240-numeric-field-bug

AnuragVasanwala committed:
Merge branch 'main' into 13238-13239-13240-numeric-field-bug

AnuragVasanwala committed:
♻️ Add check to verify `min` is `undefined`

AnuragVasanwala committed:
Merge branch 'main' into 13238-13239-13240-numeric-field-bug

AnuragVasanwala committed:
📖 Add/update necessary storybook examples

AnuragVasanwala committed:
♻️ Add missing deps and Lint

AnuragVasanwala committed:
♻️ Fix linting

AnuragVasanwala committed:
🐛 Add `updateOnChange` attribute for backward compatibility

AnuragVasanwala committed:
🐛 Remove `min` restriction

AnuragVasanwala committed:
⏪ Revert modified js test-case, fix spacing

AnuragVasanwala committed:
♻️ Add missing property `updateOnChange`

AnuragVasanwala committed:
⏪ Revert unmodified unit test case

AnuragVasanwala committed:
🧪 Fix unit test to check for invalid hour and minute

AnuragVasanwala committed:
♻️ Remove multiple assertion for single `waitFor`

swissspidy commented on packages/design-system/src/components/input/numericInput.tsx:
```diff
@@ -88,14 +91,29 @@ const NumericInput = forwardRef(function NumericInput(
+  // Holds filteredValue that may have leading padZero when required.
+  let paddedValue = '';
+  if (max !== undefined) {
+    const maxPad = max !== undefined ? String(max).length : 0;
```
```suggestion
    const maxPad = String(max).length;
```

We already checked `max !== undefined` above.

AnuragVasanwala replied:
Absolutely! I missed it. I am updating accordingly.

AnuragVasanwala committed:
🚮 Remove unnecessary `undefined` check

Co-authored-by: Pascal Birchler <pascalb@google.com>

AnuragVasanwala committed:
Merge branch 'main' into 13238-13239-13240-numeric-field-bug

AnuragVasanwala committed:
Merge branch 'main' into 13238-13239-13240-numeric-field-bug

AnuragVasanwala committed:
🐛 Fix `NaN` bug, ✨ Add support for `-` sign

AnuragVasanwala committed:
Merge branch 'main' into 13238-13239-13240-numeric-field-bug

swissspidy commented on packages/story-editor/src/components/form/dateTime/timePicker.js:
```diff
@@ -171,20 +177,20 @@ function TimePicker({ onChange, is12Hour, localeData, setLocaleData }) {
               onChange={onChangeEvent('hours')}
+              updateOnChange
               onBlur={updateHours}
-              padZero
```
Curious, why remove `padZero` again here and below?

AnuragVasanwala replied:
I intentionally removed `padZero`. I was experimenting with and without padding. It seems, we may skip padding in hours but required in minutes.

I was experimenting to mach padding with the textbox as shown below:
<img width="376" alt="image" src="https://github.com/GoogleForCreators/web-stories-wp/assets/75766877/a98659ad-235f-458e-91b5-c2fa97014018">

What do you say? Should I enable back padding for both OR just for minutes?

swissspidy replied:
> I was experimenting to mach padding with the textbox as shown below:

That is a good idea! Let's try to match padding with the time format in the textbox.

The time format is coming from Settings -> General -> Time Format.

Possible formats, as per PHP documentation:

g | Hour, 12-hour, without leading zeros | 1–12
-- | -- | --
h | Hour, 12-hour, with leading zeros | 01–12
G | Hour, 24-hour, without leading zeros | 0-23
H | Hour, 24-hour, with leading zeros | 00-23
i | Minutes, with leading zeros | 00-59

So the minutes always have leading zeros, so we can force padding there.

For the hours, `h` and `H` force leading zeros, `g` and `G` don't.

So to accommodate for that, we could introduce a new utility function similar to [`is12Hour`](https://github.com/GoogleForCreators/web-stories-wp/blob/a682e9be42e621ebfbefeedb04c40ba44ac03dc7/packages/date/src/is12Hour.ts#L22) to detect whether the hour has leading zeros or not. Maybe `hasLeadingZeros.ts`?

Then we could basically do `padZero={hasLeadingZeros()}`

What do you think? :-)

AnuragVasanwala replied:
Awesome! That's a great idea.

> So to accommodate for that, we could introduce a new utility function similar to [is12Hour](https://github.com/GoogleForCreators/web-stories-wp/blob/a682e9be42e621ebfbefeedb04c40ba44ac03dc7/packages/date/src/is12Hour.ts#L22) to detect whether the hour has leading zeros or not. Maybe `hasLeadingZeros.ts`?

Yes, `hasLeadingZeros.ts` utility sounds good 👍🏻 and shall implement accordingly.

AnuragVasanwala replied:
@swissspidy While implementing `hasLeadingZeros.ts`, I thought instead of introducing a new utility function `hasLeadingZeros`, may we use `is12Hour` like this for hour?

```diff
- padZero={hasLeadingZeros}
+ padZero={!is12Hour}
```

It seems much easy to understand. What do you suggest?

swissspidy replied:
Unfortunately not, because you can use 12 hour format both with and without leading zeros:

g | Hour, 12-hour, without leading zeros | 1–12
-- | -- | --
h | Hour, 12-hour, with leading zeros | 01–12
G | Hour, 24-hour, without leading zeros | 0-23
H | Hour, 24-hour, with leading zeros | 00-23
i | Minutes, with leading zeros | 00-59

So is12Hour alone does not tell us whether padding is needed or not

AnuragVasanwala replied:
> Unfortunately not, because you can use 12 hour format both with and without leading zeros:
> 
> g	Hour, 12-hour, without leading zeros	1–12
> h	Hour, 12-hour, with leading zeros	01–12
> G	Hour, 24-hour, without leading zeros	0-23
> H	Hour, 24-hour, with leading zeros	00-23
> i	Minutes, with leading zeros	00-59
> So is12Hour alone does not tell us whether padding is needed or not

Ohh! I missed that. I am updating accordingly.

AnuragVasanwala committed:
Merge branch 'main' into 13238-13239-13240-numeric-field-bug

AnuragVasanwala committed:
✨ Add `hasLeadingZeros` utility to determine `padZero` for hours

AnuragVasanwala committed:
🧪 Add unit test for `hasLeadingZeros`

AnuragVasanwala committed:
♻️ Fix linting issues

AnuragVasanwala committed:
♻️ Fix linting issues

AnuragVasanwala committed:
Merge branch 'main' into 13238-13239-13240-numeric-field-bug

AnuragVasanwala committed:
⏪ Remove unnecessary test case

AnuragVasanwala committed:
♻️ `allowEmpty` only when Height textbox is disabled

AnuragVasanwala committed:
🐛 Fix `NumericInput` component bug

+ Fixes bug where negative sign is not allowed
+ Fixes bug where `0` is not entered by default when `allowEmpty` is unset
+ Fixes bug when `allowEmpty` is set and empty value is not allowed

swissspidy commented on packages/design-system/src/components/input/useNumericInput.tsx:
```diff
@@ -58,20 +60,36 @@ const useNumericInput = ({
+        onChange(ev, '');
       }
     },
     [currentValue, onChange, options]
```
```suggestion
    [currentValue, onChange, options, allowEmpty, min]
```

To fix the eslint error

AnuragVasanwala replied:
Fixed in the latest commit: https://github.com/GoogleForCreators/web-stories-wp/pull/13339/commits/ea68d356ef699d864f2a6ee752341e97ab4b9a77

AnuragVasanwala committed:
🐛 Fix missing dependencies and fix a lint issue

swissspidy commented on packages/story-editor/src/components/panels/design/sizePosition/sizePosition.js:
```diff
@@ -323,6 +323,7 @@ function SizePositionPanel(props) {
         </Area>
         <Area area="h">
           <NumericInput
+            allowEmpty={disableHeight ? true : false}
```
```suggestion
            allowEmpty={disableHeight}
```

AnuragVasanwala committed:
♻️ Remove unnecessary ternary operator

Co-authored-by: Pascal Birchler <pascalb@google.com>

AnuragVasanwala committed:
Merge branch 'main' into 13238-13239-13240-numeric-field-bug

AnuragVasanwala committed:
Merge branch 'main' into 13238-13239-13240-numeric-field-bug

AnuragVasanwala committed:
Merge branch 'main' into 13238-13239-13240-numeric-field-bug

AnuragVasanwala committed:
♻️ Update to last valid number on empty string

AnuragVasanwala committed:
♻️ Fix comment spacing lint error
